### PR TITLE
Bug 1222132 - Pushing and pulling files via /files

### PR DESCRIPTION
**Pull request by eliperelman**

### CODE REVIEW

File: src/routes/files.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+let express = require('express');
+let fxos = require('../fxos');
+let multer = require('multer');
+let path = require('path');
+
+let router = express.Router();
+let storage = multer({storage: multer.diskStorage({filename})});
```
Comment (gaye) [inline src/routes/files.js]: `filename` is not defined here?

File: src/routes/files.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    res.status(422).send('Missing filepath parameter');
+  } else {
+    next();
+  }
+}
+
+function uploader(req, res, next) {
+  middleware(req, res, err => {
+    if (err) {
+      res.status(500).send('File upload failed');
```
Comment (gaye) [inline src/routes/files.js]: nit: `return` instead of else blocks to reduce nesting?

File: test/commands/shell.js [hunk 3]
```diff
@@ -23,12 +23,23 @@ function stream(command, location) {
...
+  return child;
 }
 
 var commands = {};
 
 commands.cat = function cat(file) {
-  return stream('cat', file);
+  var proc = stream('cat', file);
+
+  proc.stderr.once('data', function(data) {
```
Comment (gaye) [inline test/commands/shell.js]: nit: prefer arrow fn

File: test/commands/shell.js [hunk 4]
```diff
@@ -23,12 +23,23 @@ function stream(command, location) {
...
+    if (/No such file or directory/.test(data.toString())) {
+      console.log(data.toString());
```
Comment (gaye) [inline test/commands/shell.js]: s/log/error/

File: test/commands/shell.js [hunk 5]
```diff
@@ -67,16 +78,18 @@ commands.kill = noop;
 commands.log = noop;
 commands.setprop = noop;
 
-module.exports = function shell(arg0) {
-  if (!arg0) {
+module.exports = function shell() {
+  var args = Array.prototype.slice.call(arguments).join(' ');
```
Comment (gaye) [inline test/commands/shell.js]: Spread operator?

Comment (gaye) [inline test/commands/shell.js]: Oh right this is es5 lol

File: test/commands/shell.js [hunk 6]
```diff
@@ -67,16 +78,18 @@ commands.kill = noop;
...
+
+  if (!args) {
     return;
   }
 
-  if (arg0.indexOf("'") === 0 && arg0.indexOf("'") === 0) {
-    arg0 = arg0.slice(1, -1);
+  if (args.indexOf("'") === 0 && args.indexOf("'") === 0) {
```
Comment (gaye) [inline test/commands/shell.js]: This is the same predicate twice right?

File: test/routes/files_test.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
+let get = require('../get');
```
Comment (gaye) [inline test/routes/files_test.js]: nit: f before g :)

File: test/routes/files_test.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+suite('PUT /files', () => {
+  let uploadFile;
+
+  setup(async function() {
+    uploadFile = await file(path.resolve(`${__dirname}/../fixtures/profiles.ini`));
+  });
+
+  test('should upload OK', async function() {
+    let url = 'http://localhost:3000/files?filepath=/data/b2g/mozilla/profiles.ini';
```
Comment (gaye) [inline test/routes/files_test.js]: Do we need to urlencode the filepaths?

**Comment by gaye** [comment]
Overall this one looks really good -- have a few nits but nothing major afaict

File: src/routes/files.js (hunk 2 above)
Comment (eliperelman) [inline src/routes/files.js]: :+1:

File: test/commands/shell.js (hunk 4 above)
Comment (eliperelman) [inline test/commands/shell.js]: Current adb emits this on stdout, which is why I also have it going there. If you look at routes/files.js (currently line 41), you'll see that real adb for some reason pushes this on stdout, and I commented it as such as well.

File: test/commands/shell.js (hunk 3 above)
Comment (eliperelman) [inline test/commands/shell.js]: As noted later, fake adb is all ES5.

File: test/commands/shell.js (hunk 5 above)
Comment (eliperelman) [inline test/commands/shell.js]: :smile:

File: test/commands/shell.js (hunk 6 above)
Comment (eliperelman) [inline test/commands/shell.js]: wtf.

This came from my refactor to ES5 and I removed `startsWith` and `endsWith` and, well, fail.

File: test/routes/files_test.js (hunk 7 above)
Comment (eliperelman) [inline test/routes/files_test.js]: Derp, how do I alphabet?

File: test/routes/files_test.js (hunk 8 above)
Comment (eliperelman) [inline test/routes/files_test.js]: __We__ probably don't need to in the test, but consumers should probably be URL encoding all their stuff. Right now I don't see anything that needs URL encoding. Thoughts?

File: test/routes/files_test.js (hunk 7 above)
Comment (eliperelman) [inline test/routes/files_test.js]: Fixed.

File: test/commands/shell.js (hunk 3 above)
Comment (gaye) [inline test/commands/shell.js]: yep yep my bad!

File: test/commands/shell.js (hunk 4 above)
Comment (gaye) [inline test/commands/shell.js]: Ah okay fair enough

File: test/routes/files_test.js (hunk 8 above)
Comment (gaye) [inline test/routes/files_test.js]: Okay sounds fine to me

**Pull request merged**